<p>Python has no pre/post increment/decrement operator. For instance, <code>x++</code> and <code>x--</code> will fail to parse. More importantly,
<code>++x</code> and <code>--x</code> will do nothing. To increment a number, simply write <code>x += 1</code>.</p>
<h2>Noncompliant Code Example</h2>
<pre>
++x # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
x += 1
</pre>

