<p>If a local variable is declared but not used, it is dead code and should be removed. Doing so will improve maintainability because developers will
not wonder what the variable is used for.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def hello(name):
    message = "Hello " + name # Noncompliant
    print(name)
</pre>
<h2>Compliant Solution</h2>
<pre>
def hello(name):
    message = "Hello " + name
    print(message)
</pre>

